File: problems/2026/03/2026-03-26-leetcode-3-无重复字符最长子串/SolutionLC3.java
import java.util.HashMap;
import java.util.Map;

public class SolutionLC3 {

    public int lengthOfLongestSubstring(String s) {
        // 记录字符最近一次出现的位置：字符 -> 下标
        Map<Character, Integer> lastSeen = new HashMap<>();

        // left 表示当前滑动窗口左边界，maxLen 记录历史最大长度
        int left = 0;
        int maxLen = 0;

        for (int right = 0; right < s.length(); right++) {
            char current = s.charAt(right);

            // 如果当前字符在窗口中已经出现过，需要收缩左边界
            if (lastSeen.containsKey(current) && lastSeen.get(current) >= left) {
                // 跳到重复字符上次出现位置的后一位，保证窗口内无重复
                left = lastSeen.get(current) + 1;
            }

            // 更新当前字符最新出现位置
            lastSeen.put(current, right);

            // 用当前窗口长度更新答案
            int currentLen = right - left + 1;
            maxLen = Math.max(maxLen, currentLen);
        }

        return maxLen;
    }

    public static void main(String[] args) {
        SolutionLC3 solution = new SolutionLC3();

        String[] inputs = {"abcabcbb", "bbbbb", "pwwkew", "", "dvdf"};
        int[] expected = {3, 1, 3, 0, 3};

        for (int i = 0; i < inputs.length; i++) {
            int result = solution.lengthOfLongestSubstring(inputs[i]);
            System.out.println("调用: lengthOfLongestSubstring(\"" + inputs[i] + "\")");
            System.out.println("预期结果: " + expected[i]);
            System.out.println("实际结果: " + result);
            System.out.println();
        }
    }
}

File: problems/2026/03/2026-03-26-leetcode-3-无重复字符最长子串/SolutionLC3.py
class Solution:
    def length_of_longest_substring(self, s: str) -> int:
        # 记录字符最近一次出现的位置：字符 -> 下标
        last_seen = {}

        # left 是窗口左边界，max_len 是最终答案
        left = 0
        max_len = 0

        for right, ch in enumerate(s):
            # 如果字符在当前窗口内重复出现，需要移动左边界
            if ch in last_seen and last_seen[ch] >= left:
                # 直接跳到重复字符上次位置后一位
                left = last_seen[ch] + 1

            # 更新当前字符的最新位置
            last_seen[ch] = right

            # 计算当前无重复窗口长度并更新答案
            current_len = right - left + 1
            if current_len > max_len:
                max_len = current_len

        return max_len


if __name__ == "__main__":
    solution = Solution()

    inputs = ["abcabcbb", "bbbbb", "pwwkew", "", "dvdf"]
    expected = [3, 1, 3, 0, 3]

    for i in range(len(inputs)):
        result = solution.length_of_longest_substring(inputs[i])
        print(f'调用: length_of_longest_substring("{inputs[i]}")')
        print(f"预期结果: {expected[i]}")
        print(f"实际结果: {result}")
        print()
